tools/wptrunner/wptrunner/executors/protocol.py

```diff
@@ -393,29 +393,46 @@ def switch_to_window(self, wptrunner_id, initial_window=None):
                 if not first or item != initial_window:
                     self.parent.base.set_window(item)
             else:
-                self._switch_to_frame(item)
+                try:
+                    self._switch_to_frame(item)
+                except ValueError:
+                    # The frame no longer exists, or doesn't have a nested browsng context, so continue
+                    pass
 
             try:
-                handle_window_id = self.parent.base.execute_script("return window.__wptrunner_id")
-                if str(handle_window_id) == wptrunner_id:
-                    return
+                # Get the window id and a list of elements containing nested browsing contexts.
+                # For embed we can't tell fpr sure if there's a nested browsing context, so always return it
+                # and fail later if there isn't
+                result = self.parent.base.execute_script("""
+                let contextParents = Array.from(document.querySelectorAll("frame, iframe, embed, object"))
+                    .filter(elem => elem.localName !== "embed" ? (elem.contentWindow !== null) : true);
+                return [window.__wptrunner_id, contextParents]""")
             except Exception:
-                pass
-            frame_count = self.parent.base.execute_script("return window.length")
-            if frame_count:
-                for frame_id in reversed(range(0, frame_count)):
-                    # None here makes us switch back to the parent after we've processed the frame
-                    stack.append(None)
-                    stack.append(frame_id)
+                continue
+
+            if result is None:
+                # With marionette at least this is possible if the content process crashed. Not quite
+                # sure how we want to handle that case.
+                continue
+
+            handle_window_id, nested_context_containers = result
+
+            if handle_window_id and str(handle_window_id) == wptrunner_id:
+                return
+
+            for elem in reversed(nested_context_containers):
+                # None here makes us switch back to the parent after we've processed the frame
+                stack.append(None)
+                stack.append(elem)
             first = False
```

Comment:
The new `continue`s skip this line that clears `first`, which can lead to problems if the window has been changed and the next is `initial_window`.  Perhaps it's simplest to clear `first` immediately after `self.parent.base.set_window(item)`.  Or `first` could be cleared immediately before the `execute_script` `try` block, which would in some ways be more clearly correct.

Reply:
Oops. Good catch.